_posts/2016-07-11-value-of-automated-e2e-testing.md

```diff
@@ -0,0 +1,58 @@
+
+However, despite the impediments, these tests were found very helpful. The object was reached, the tests checked that the icons were visible and clickable in the specified situations, and even more, if you broke functionality by accident it was exposed by a test.
+
+##The New Era with JBehave and Selenium
+
+After a while, we got a testing specialist in our project, and a new era began. He started to build real automated e2e-tests with JBehave and Selenium. JBehave enables writing the tests as stories in text files. Each story file has a matching Java class implementing the steps described in the story. The steps and the actual test code are mapped together with annotations. The web page actions, button clicks, element searches and so on are written in separate bridge classes (e.g. MainItemPageBridge.java). The tests are run with the help of Selenium Web Driver, which links the actual web page and the test code together.
+
+![An example of JBehave and Selenium in action](/img/value-of-e2e-testing/jbehave.png)
+
+At first, there was a huge gap between code implemented and code tested with JBehave tests. There were a couple of developers just writing tests for the already implemented features. However, soon it was clear, that the gap would never be surpassed this way. Everyone in the project should learn writing the new type of tests. In my opinion, in the beginning it was slow and not very easy but step by step it became a natural way to write tests. Of course, the test framework was developed side by side with the tests, because we wrote a lot of common utility methods and tools.
+
+The JBehave tests were slower to run than Wicket tests, because opening the browser took time. When the number of JBehave tests grew, we realized that running all the e2e tests on our own workstations was too time consuming. Hudson, which is our CI platform, was configured to run all the tests after a bunch of commits. This way we could run only the relevant tests on our own workstations and the full regression tests were run by Hudson. Hudson is also configured to save the print screen images if JBehave tests failed. The images are a huge help when trying to figure out why the test failed. The inspected element might be missing or maybe the test is not clicking the right element at all. The JBehave tests are also worthwhile when debugging the application. The baseline can be set up programmatically before the browser is started. After that, you can run the test step by step watching similarly the actual action on the web page.
```

Comment:
There are sentences both in past and present tense. Most are in past tense except "Hudson, which is.."